gtsam/basis/ParameterMatrix.h

```diff
@@ -0,0 +1,218 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file ParamaterMatrix.h
+ * @brief Define ParameterMatrix class which is used to store values at
+ * interpolation points.
+ * @author Varun Agrawal, Frank Dellaert
+ * @date September 21, 2020
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Testable.h>
+#include <gtsam/base/VectorSpace.h>
+
+#include <iostream>
+
+namespace gtsam {
+
+/**
+ * A matrix abstraction of MxN values at the Basis points.
+ * This class serves as a wrapper over an Eigen matrix.
+ * @tparam T: the type you would like to evaluate using polynomial
```

Comment:
is `T` supposed to be `M` ???

Also can you give a quick reason this is necessary over just normal eigen matrices?  Is it for type-safety reasons?

Reply:
The idea is to have a matrix where the rows define the dimensionality of the problem (e.g. `ParameterMatrix<traits<STATE>::dimension>`) but whose columns are variable in number and can be adjusted as per the value of `N`.

Reply:
Hmm, still is there a reason to not just
```c++
template <int M>
using ParameterMatrix = Eigen::Matrix<double, M, -1>
```
?

Reply:
I honestly don't remember. There was a reason when I first implemented it. I'll have to go back and look at doubleExpresso to find the reason.

Reply:
I remembered the reason.
Doing 
```cpp
template <int M>
using ParameterMatrix = Eigen::Matrix<double, M, -1>
```
causes the API to become far more verbose and non-intuitive.

E.g., we can no longer do `ParameterMatrix<12> X(N)` since it gives an error that we tried calling a vector method on a matrix object. We have to instead do
```cpp
ParameterMatrix<12> X;
X.resize(12, N); // why do we need to specify 12 again here? This is something eigen needs :(
```

In general, the fact that one of the dimensions is dynamic and the other has size greater than 1 makes things non-standard for Eigen since it assumes  everything is a Matrix and simple things like `X.col` don't compile.

I'll add this to the docstring for future reference.

Reply:
Ahh got it - makes sense!  Thanks!